### PR TITLE
move post.viewcounts to helper.py and cache with dogpile

### PR DESCRIPTION
**Pull request by shreyas-satish**

### CODE REVIEW

File: hasjob/templates/detail.html [hunk 1]
```diff
@@ -231,7 +231,7 @@
             {%- endif %}
             {%- if is_siteadmin or post.admin_is(g.user) or (g.user and g.user.flags.is_employer_month) %}
               <div class="list-group-item no-border">
-                <i class="fa fa-fw fa-bar-chart-o"></i>&nbsp;&nbsp; <span title="Viewed › Opened form › Applied">{{ post.viewcounts['viewed'] }} › {{ post.viewcounts['opened'] }} › {{ post.viewcounts['applied'] }}</span> &middot;
+                <i class="fa fa-fw fa-bar-chart-o"></i>&nbsp;&nbsp; <span title="Viewed › Opened form › Applied">{{ get_post_viewcounts(post)['viewed'] }} › {{ get_post_viewcounts(post)['opened'] }} › {{ get_post_viewcounts(post)['applied'] }}</span> &middot;
```
Comment (jace) [inline hasjob/templates/detail.html]: This hits cache thrice. It should be wrapped in a single `with` statement. Ideally for all uses within the template.

File: hasjob/views/helper.py [hunk 2]
```diff
@@ -489,7 +545,7 @@ def list_dirty_impression_counts():
 
 def update_impression_counts(jobpost_ids):
     for jobpost_id in jobpost_ids:
-        # Replicate post.viewcounts's query here because we don't want to load the post from db just to
+        # Replicate viewcounts query from JobPost here because we don't want to load the post from db just to
```
Comment (jace) [inline hasjob/views/helper.py]: This excuse no longer applies as it's not in the JobPost class. The call could be abstracted into a separate function.

File: hasjob/views/index.py [hunk 3]
```diff
@@ -33,13 +33,13 @@ def stickie_dict(post, url, pinned=False, show_viewcounts=False, show_pay=False,
...
     if show_viewcounts:
         result['viewcounts'] = {
-            'listed': post.viewcounts['impressions'],
-            'viewed': post.viewcounts['viewed'],
-            'opened': post.viewcounts['opened'],
-            'applied': post.viewcounts['applied']
+            'listed': get_post_viewcounts(post)['impressions'],
+            'viewed': get_post_viewcounts(post)['viewed'],
+            'opened': get_post_viewcounts(post)['opened'],
+            'applied': get_post_viewcounts(post)['applied']
```
Comment (jace) [inline hasjob/views/index.py]: You're no longer using `@cached_property` so this will hit cache multiple times. Call it once and extract values.

File: hasjob/views/index.py [hunk 4]
```diff
@@ -404,6 +404,8 @@ def index(basequery=None, md5sum=None, tag=None, domain=None, location=None, tit
     data['now'] = now
     data['is_siteadmin'] = is_siteadmin
     data['location_prompts'] = location_prompts
+    if data['domain'] and data['domain'] not in db.session:
+        data['domain'] = db.session.merge(data['domain'])
```
Comment (jace) [inline hasjob/views/index.py]: Can you move these lines into a separate commit and push to master? It shouldn't be in this PR.

File: hasjob/views/listing.py [hunk 5]
```diff
@@ -88,7 +88,8 @@ def jobdetail(domain, hashid):
                 db.session.commit()
             except IntegrityError:
                 db.session.rollback()
-            post.viewcounts  # Re-populate cache
+            # Re-populate cache
+            get_post_viewcounts(post)
```
Comment (jace) [inline hasjob/views/listing.py]: This won't work. Tell Dogpile to clear instead.

File: hasjob/views/listing.py [hunk 6]
```diff
@@ -243,7 +241,8 @@ def revealjob(domain, hashid):
             cache.delete_memoized(viewstats_by_id_qhour, post.id)
             cache.delete_memoized(viewstats_by_id_hour, post.id)
             cache.delete_memoized(viewstats_by_id_day, post.id)
-            post.viewcounts  # Re-populate cache
+            # Re-populate cache
+            get_post_viewcounts(post)
```
Comment (jace) [inline hasjob/views/listing.py]: Ask Dogpile instead.

File: hasjob/views/listing.py [hunk 7]
```diff
@@ -254,7 +253,8 @@ def revealjob(domain, hashid):
         cache.delete_memoized(viewstats_by_id_qhour, post.id)
         cache.delete_memoized(viewstats_by_id_hour, post.id)
         cache.delete_memoized(viewstats_by_id_day, post.id)
-        post.viewcounts  # Re-populate cache
+        # Re-populate cache
+        get_post_viewcounts(post)
```
Comment (jace) [inline hasjob/views/listing.py]: Dogpile.

File: hasjob/templates/detail.html (hunk 1 above)
Comment (shreyas-satish) [inline hasjob/templates/detail.html]: Fixed. Now calls `get_post_viewcounts` in the view and passes it on to the template.

File: hasjob/views/helper.py [hunk 8]
```diff
@@ -320,6 +320,62 @@ def bgroup(jobpost_ab, post):
...
+        redis_store.hset(cache_key, 'pay_label', values['pay_label'].encode('utf-8'))
+        redis_store.expire(cache_key, 86400)
+    elif isinstance(values['pay_label'], str):  # Redis appears to return bytestrings, not Unicode
+        values['pay_label'] = unicode(values['pay_label'], 'utf-8')
+    return values
+
+
+@app.context_processor
+def inject_post_viewcounts():
+    return dict(get_post_viewcounts=get_post_viewcounts)
```
Comment (shreyas-satish) [inline hasjob/views/helper.py]: Fixed.

File: hasjob/views/index.py (hunk 3 above)
Comment (shreyas-satish) [inline hasjob/views/index.py]: Fixed.

File: hasjob/views/index.py (hunk 4 above)
Comment (shreyas-satish) [inline hasjob/views/index.py]: Done.

File: hasjob/views/listing.py (hunk 5 above)
Comment (shreyas-satish) [inline hasjob/views/listing.py]: Done.

File: hasjob/views/listing.py (hunk 6 above)
Comment (shreyas-satish) [inline hasjob/views/listing.py]: Done.

File: hasjob/views/listing.py (hunk 7 above)
Comment (shreyas-satish) [inline hasjob/views/listing.py]: Done.

File: hasjob/models/jobpost.py [hunk 9]
```diff
@@ -205,6 +204,12 @@ def get(cls, hashid):
         return cls.query.filter_by(hashid=hashid).one_or_none()
 
     @classmethod
+    def get_by_id(cls, jobpost_id, load_id_only=False):
```
Comment (jace) [inline hasjob/models/jobpost.py]: This function is unnecessary. You can use `JobPost.query.get(<id>)`

File: hasjob/models/jobpost.py [hunk 10]
```diff
@@ -426,59 +431,13 @@ def tag_content(self):
...
-            redis_store.hset(cache_key, 'pay_label', values['pay_label'].encode('utf-8'))
-            redis_store.expire(cache_key, 86400)
-        elif isinstance(values['pay_label'], str):  # Redis appears to return bytestrings, not Unicode
-            values['pay_label'] = unicode(values['pay_label'], 'utf-8')
-        return values
+        return 'hasjob/viewcounts/%d' % jobpost_id
 
     def uncache_viewcounts(self, key=None):
-        cache_key = self.viewcounts_key
+        cache_key = self.viewcounts_key(self.id)
```
Comment (jace) [inline hasjob/models/jobpost.py]: Use `JobPost.viewcounts_key` when calling a static method. This is very confusing. Alternatively, move this function out of the class and have the property call the external function.

File: hasjob/templates/macros.html [hunk 11]
```diff
@@ -8,7 +8,19 @@
     <span class="annotation top-left">{{ post.location }}</span>
     <span class="annotation top-right">{{ post.datetime|shortdate }}</span>
     <span class="headline">{% if is_bgroup %}{{ post.headlineb }}{% else %}{{ post.headline }}{% endif %}</span>
-    {% if show_viewcounts or show_pay %}<span class="count">{% if show_viewcounts %}<span title="Listed › Viewed › Opened form › Applied">{{ post.viewcounts['impressions'] }} › {{ post.viewcounts['viewed'] }} › {{ post.viewcounts['opened'] }} › {{ post.viewcounts['applied'] }}</span>{% endif %}{% if show_viewcounts and show_pay %} &middot; {% endif %}{% if show_pay %}{{ post.viewcounts['pay_label'] }}{% endif %}</span>{% endif %}
+    {% if show_viewcounts or show_pay %}
+      {% with post_viewcounts=get_post_viewcounts(post.id) %}
```
Comment (jace) [inline hasjob/templates/macros.html]: Use `{%-` as this will add a blank line.

File: hasjob/templates/macros.html [hunk 12]
```diff
@@ -8,7 +8,19 @@
...
+      <span class="count">
+        {% if show_viewcounts %}
```
Comment (jace) [inline hasjob/templates/macros.html]: Likewise for all template tags.

File: hasjob/views/helper.py [hunk 13]
```diff
@@ -320,9 +320,68 @@ def bgroup(jobpost_ab, post):
     return (jobpost_ab.get(post.id) or cointoss()) if post.headlineb else None
 
 
-def cache_viewcounts(posts):
+def get_jobpost_impressions(jobpost_id):
+    return db.session.query(db.func.count(db.func.distinct(EventSession.user_id)).label('count')).join(
+        JobImpression).filter(JobImpression.jobpost_id == jobpost_id).first().count
```
Comment (jace) [inline hasjob/views/helper.py]: Replace `.first().count` with `.scalar()`

File: hasjob/views/helper.py [hunk 14]
```diff
@@ -320,9 +320,68 @@ def bgroup(jobpost_ab, post):
...
+
+
+def get_post_viewcounts(jobpost_id):
+    cache_key = JobPost.viewcounts_key(jobpost_id)
+    values = g.viewcounts.get(cache_key)
+    jobpost = None
+    if not values:
+        values = redis_store.hgetall(cache_key)
+        jobpost = JobPost.get_by_id(jobpost_id)
```
Comment (jace) [inline hasjob/views/helper.py]: `jobpost = JobPost.query.get(jobpost_id)`

File: hasjob/views/helper.py [hunk 15]
```diff
@@ -320,9 +320,68 @@ def bgroup(jobpost_ab, post):
...
+    return db.session.query(db.func.count(db.func.distinct(EventSession.user_id)).label('count')).join(
+        JobImpression).filter(JobImpression.jobpost_id == jobpost_id).first().count
+
+
+def get_post_viewcounts(jobpost_id):
+    cache_key = JobPost.viewcounts_key(jobpost_id)
+    values = g.viewcounts.get(cache_key)
+    jobpost = None
+    if not values:
+        values = redis_store.hgetall(cache_key)
```
Comment (jace) [inline hasjob/views/helper.py]: Are you sure `values` is a dict even when there is no key in Redis? If this is confirmed behaviour, please leave a comment to the effect.

Comment (jace) [inline hasjob/views/helper.py]: It is indeed always a dict: https://github.com/andymccurdy/redis-py/blob/master/redis/client.py#L406

Please indicate so in a comment.

File: hasjob/views/helper.py (hunk 14 above)
Comment (jace) [inline hasjob/views/helper.py]: Also do this in the following few calls.

File: hasjob/views/helper.py [hunk 16]
```diff
@@ -320,9 +320,68 @@ def bgroup(jobpost_ab, post):
...
+
+@app.context_processor
+def inject_post_viewcounts():
+    return {'get_post_viewcounts': get_post_viewcounts}
+
+
+def load_viewcounts(posts):
     redis_pipe = redis_store.pipeline()
-    viewcounts_keys = [p.viewcounts_key for p in posts]
+    viewcounts_keys = [JobPost.viewcounts_key(p.id) for p in posts]
```
Comment (jace) [inline hasjob/views/helper.py]: This is ~700 calls in a row in production. That slowdown begins to be noticeable. Make `viewcounts_key` accept a list and return a list (test with `isinstance(jobpost_id, (list, tuple))`)

File: hasjob/views/helper.py [hunk 17]
```diff
@@ -489,11 +548,9 @@ def list_dirty_impression_counts():
 
 def update_impression_counts(jobpost_ids):
     for jobpost_id in jobpost_ids:
-        # Replicate post.viewcounts's query here because we don't want to load the post from db just to
-        # access a class instance method
-        redis_store.hset('hasjob/viewcounts/%d' % jobpost_id, 'impressions',
-            db.session.query(db.func.count(db.func.distinct(EventSession.user_id)).label('count')
-            ).join(JobImpression).filter(JobImpression.jobpost_id == jobpost_id).first().count)
+        redis_store.hset('hasjob/viewcounts/%d' % jobpost_id,
```
Comment (jace) [inline hasjob/views/helper.py]: You're replicating the key syntax here but (a) not mentioning the existing `JobPost.viewcounts_key` staticmethod, nor (b) using it. Pick one.

Comment (shreyas-satish) [inline hasjob/views/helper.py]: Fixed.

File: hasjob/views/helper.py (hunk 16 above)
Comment (shreyas-satish) [inline hasjob/views/helper.py]: Fixed.

File: hasjob/views/helper.py (hunk 14 above)
Comment (shreyas-satish) [inline hasjob/views/helper.py]: Fixed.

File: hasjob/views/helper.py (hunk 15 above)
Comment (shreyas-satish) [inline hasjob/views/helper.py]: Fixed.

File: hasjob/views/helper.py (hunk 13 above)
Comment (shreyas-satish) [inline hasjob/views/helper.py]: Fixed.

File: hasjob/templates/macros.html (hunk 12 above)
Comment (shreyas-satish) [inline hasjob/templates/macros.html]: Fixed.

File: hasjob/templates/macros.html (hunk 11 above)
Comment (shreyas-satish) [inline hasjob/templates/macros.html]: Fixed.

File: hasjob/models/jobpost.py (hunk 9 above)
Comment (shreyas-satish) [inline hasjob/models/jobpost.py]: Since `JobPost.query.get(<id>)` loads all the columns and associated relationships, I added this method to enable a more efficient fetch. I've removed the method now.

File: hasjob/templates/macros.html [hunk 18]
```diff
@@ -8,10 +8,22 @@
...
+        {%- if show_viewcounts %}
+          <span title="Listed › Viewed › Opened form › Applied">
+            {{ post_viewcounts['impressions'] }} › {{ post_viewcounts['viewed'] }} › {{ post_viewcounts['opened'] }} › {{ post_viewcounts['applied'] }}
+          </span>
+        {% endif %}
+        {%- if show_viewcounts and show_pay %} &middot; {% endif %}
+        {%- if show_pay %}{{ post_viewcounts['pay_label'] }}{% endif %}
+      </span>
+      {% endwith %}
+    {% endif %}
```
Comment (jace) [inline hasjob/templates/macros.html]: `{%-` for _all_ these template tags that are on their own line, please.

File: hasjob/templates/macros.html [hunk 19]
```diff
@@ -8,10 +8,22 @@
...
     <span class="annotation bottom-right">{{ post.company_name }}</span>
     <span class="annotation bottom-left">
-      {% if starred is not none %}
+      {%- if starred is not none %}
         <span class="pstar"><i class="fa fa-star{% if not starred %}-o{% endif %}" data-id="{{ post.hashid }}"></i></span>
       {% endif %}
```
Comment (jace) [inline hasjob/templates/macros.html]: Here too.

File: hasjob/templates/macros.html (hunk 18 above)
Comment (shreyas-satish) [inline hasjob/templates/macros.html]: Fixed.

File: hasjob/templates/macros.html (hunk 19 above)
Comment (shreyas-satish) [inline hasjob/templates/macros.html]: Fixed.
